hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/DelegatingFileSystem.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.crypto2.hadoop;
+
+import java.io.IOException;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.FileUtil;
+import org.apache.hadoop.fs.FilterFileSystem;
+import org.apache.hadoop.fs.Path;
+
+/**
+ * Equivalent to {@link FilterFileSystem} but invokes {@link #create} and {@link #open} on this class when calling
+ * {@link #copyFromLocalFile} and {@link #copyToLocalFile}.
+ * <p>
+ * Solves: https://issues.apache.org/jira/browse/HADOOP-13870
+ */
+public abstract class DelegatingFileSystem extends FilterFileSystem {
```

Comment:
should/could we not use FilterFileSystem at all?

Reply:
we'd end up reimplementing the whole thing since we do want to delegate every other method

Reply:
yeah, I know. just wondering how stable that FFS is over different Hadoop versions? do we rely on unsafe subclassing in a way that would bite us?

Reply:
think since it's just delegating we don't need to worry too much.